docs/internals/data_file.md

```diff
@@ -0,0 +1,182 @@
+atomically by swapping the pointer to the root node.
+
+Superblock is what holds this logical "root pointer" (physically, the "root pointer" is actually
+many block references). Superblock is located at a fixed position in the data file, so, when a
+replica starts up, it can read the superblock, read root block indexes and hashes from the
+superblock, and through those get access to the rest of the data in the grid. Besides the logical
+root, superblock also stores a compressed bitset of all grid blocks which are not currently
+allocated.
+
+```zig
+pub const SuperBlock = struct {
+    root: BlockReference[],
+    free_set: BitSet,
+};
+```
+
+Superblock is updated atomically and relatively infrequently. So, the normal mode of operation is
+that the replica starts up, reads the current superblock and free set to memory, then proceeds
+allocating and writing new grid blocks, picking up free entries from the bit set. That is, although
+the replica does write freshly allocated grid blocks to disk immediately, it does not update the
+superblock on disk (so the logical state reachable from the superblock stays the same). Only after a
+relatively large amount of new grid blocks written, replica atomically writes the new superblock,
+with a new free set and a new logical "root pointer". To implement atomic update of the superblock,
+the superblock is physically stored as 4 distinct copies on disk. After startup, replica picks the
+latest superblock which has at least 2 copies written. Picking just the latest copy would be wrong
+--- unlike the grid blocks, the superblock stores its own checksum, and is vulnerable to misdirected
+reads (i.e., a misdirected read can hide the sole latest copy).
+
+Because the superblock (and hence, logical grid state) is updated infrequently and in bursts, it
+can't represent the entirety of persistent state. The rest of the state is stored in the write ahead
+log. WAL is a ring buffer with prepares, and represents the logical diff which should be applied to
+the state represented by superblock/grid to get the actual current state of the system. When a
+replica processes a prepare it, roughly:
+
+* writes it to WAL on disk
+* applies changes from prepare to the in-memory data structure representing the current state
+* applies changes from the prepare to the pending on-disk state by allocating and writing fresh grid
+  blocks
+
+When enough prepares are received, the superblock is updated to point the accumulated-so-far new
+disk state.
+
+This covers how the three major areas of the data file -- the write-ahead log, the superblock and
+the grid -- work together to represent abstract persistent logical state.
+
+Concretely, the sate of TigerBeetle is a collection (forest) of LSM trees. LSM structure is
+described [in a separate document](./lsm.md), here only high level on-disk layout is discussed.
+
+Each LSM tree stores a set of values. Values are:
+
+* fixed in size,
```

Comment:
Maybe s/fixed/uniform? i.e. all values in a particular LSM tree have the same size.
("Fixed" implies that the size doesn't change over time, which is also true, but maybe less surprising.)